### PR TITLE
Log shopfloor requests in a table

### PR DESCRIPTION
**Pull request by guewen**
Selection_853
Selection_855
Selection_854

card 809

### CODE REVIEW

File: shopfloor/readme/CONFIGURE.rst [hunk 1]
```diff
@@ -1 +1,10 @@
 writeme
+
+Logs retention
+--------------
```
Comment (simahawk) [inline shopfloor/readme/CONFIGURE.rst:4]: name what is logged

File: shopfloor/services/service.py [hunk 2]
```diff
@@ -23,6 +24,51 @@ class BaseShopfloorService(AbstractComponent):
     _actions_collection_name = "shopfloor.action"
     _expose_model = None
 
+    def dispatch(self, method_name, _id=None, params=None):
```
Comment (simahawk) [inline shopfloor/services/service.py:29]: add a TODO: this should be splitted to an extra module for rest-framework

File: shopfloor/services/service.py [hunk 3]
```diff
@@ -23,6 +24,51 @@ class BaseShopfloorService(AbstractComponent):
...
+                self._log_call_in_db(env, _id, params, error=err)
+            raise
+        self._log_call_in_db(self.env, _id, params, result=result)
+        return result
+
+    @property
+    def _log_call_header_strip(self):
+        return ("Cookie", "Api-Key")
+
+    def _log_call_in_db_values(self, _id, params, result=None, error=None):
```
Comment (simahawk) [inline shopfloor/services/service.py]: I'd pass request as an arg to make it easier to test

File: shopfloor/services/service.py [hunk 4]
```diff
@@ -23,6 +24,51 @@ class BaseShopfloorService(AbstractComponent):
...
+    def dispatch(self, method_name, _id=None, params=None):
+        try:
+            result = super().dispatch(method_name, _id=_id, params=params)
+        except Exception as err:
+            self.env.cr.rollback()
+            with registry(self.env.cr.dbname).cursor() as cr:
+                env = self.env(cr=cr)
+                self._log_call_in_db(env, _id, params, error=err)
+            raise
+        self._log_call_in_db(self.env, _id, params, result=result)
```
Comment (simahawk) [inline shopfloor/services/service.py]: isn't this too much? Can't we make it only for specific services? I'd add a class flag to do it only where we need (eg: no reason to track logs for loading menu or profiles).

**Review by simahawk — CHANGES_REQUESTED** [review]

File: shopfloor/data/ir_cron_data.xml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo noupdate="1">
+    <record id="ir_cron_autovacuum_shopfloor_log" model="ir.cron">
+        <field name="name">AutoVacuum Shopfloor Logs</field>
```
Comment (sebalix) [inline shopfloor/data/ir_cron_data.xml]: A memory of Éric Caudal which is still in my head! https://github.com/OCA/server-tools/pull/360#discussion_r52842317
Will you honor him by fixing this label? :)

File: shopfloor/services/service.py (hunk 4 above)
Comment (guewen) [reply shopfloor/services/service.py]: Are you sure there are no reason? Knowing the user didn't refresh the menu after a manager changed the configuration is something I'd like to know the day a weird bug happens (even if today it shouldn't be an issue).

Comment (guewen) [reply shopfloor/services/service.py]: That being said, you are right it's a good idea to have a flag, I'll add it

File: shopfloor/views/shopfloor_log_views.xml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                    <group
+                        string="Error"
+                        name="error"
+                        attrs="{'invisible': [('state', '!=', 'failed')]}"
+                    >
+                        <group>
+                            <field
+                                name="error"
+                                nolabel="1"
+                                widget="CopyClipboardText"
```
Comment (sebalix) [inline shopfloor/views/shopfloor_log_views.xml:65]: Thanks, didn't know that widget :)

Comment (guewen) [reply shopfloor/views/shopfloor_log_views.xml:65]: I didn't neither :D

**Comment by guewen** [comment]
@simahawk @sebalix updated, I moved the condition to enter the `_dispatch_with_db_logging` method right in `_dispatch`, so if one of the condition that disable the logging (no request, no class flag or no logging retention), we  don't have to open a new env and cr for nothing when an exception happens.

File: shopfloor/data/ir_cron_data.xml (hunk 5 above)
Comment (guewen) [reply shopfloor/data/ir_cron_data.xml]: Done

**Review by sebalix — APPROVED** [review]

File: shopfloor/services/service.py [hunk 7]
```diff
@@ -22,6 +23,63 @@ class BaseShopfloorService(AbstractComponent):
...
+            "state": "success" if result else "failed",
+        }
+
+    def _log_call_in_db(self, env, _request, _id, params, result=None, error=None):
+        values = self._log_call_in_db_values(
+            _request, _id, params, result=result, error=error
+        )
+        if not values:
+            return
+        env["shopfloor.log"].sudo().create(values)
```
Comment (simahawk) [inline shopfloor/services/service.py:82]: TODO: use queue.job?

**Review by simahawk — APPROVED** [review]

File: shopfloor/services/service.py (hunk 7 above)
Comment (guewen) [reply shopfloor/services/service.py:82]: Inserting a new row in queue_job will cost more than inserting a new row in shopfloor_log, and as it's insert-only, we won't have concurrent transaction errors

**Pull request merged**